File: frontend/src/types/chart.ts
/**
 * 图表契约（ChartSpec）：投递页分析「自由配图」的稳定形状（2026-09-15 方向定稿）。
 *
 * analyze Agent 在产出分析叙事的同时可选配 0~2 张图。本类型与后端 `app/schemas/chart.py`
 * **字段一一对应**（双侧同构），改一边必须改另一边。前端渲染前经 `isChartSpec` 校验兜底——
 * 模型配错图时静默降级（不渲染、不崩），绝不让坏 spec 白屏。
 */

/** 开放的 6 种图型（ChartRenderer 用 @mantine/charts 一一渲染）。 */
export type ChartType = 'bar' | 'line' | 'area' | 'donut' | 'pie' | 'radar'

export interface ChartSeries {
  key: string
  label?: string
  color?: string
}

export interface ChartSpec {
  type: ChartType
  title?: string
  data: Record<string, unknown>[]
  x: string
  series: ChartSeries[]
  orientation?: 'horizontal' | 'vertical'
  note?: string
}

const KNOWN_TYPES: readonly string[] = ['bar', 'line', 'area', 'donut', 'pie', 'radar']

/**
 * 运行时校验：这份 spec 能不能画（模型可能配错——类型不认得 / data 空 / x、series.key
 * 对不上 data 行）。任一不过 → false，调用方降级（跳过这张图，不渲染、不崩）。
 */
export function isChartSpec(raw: unknown): raw is ChartSpec {
  if (typeof raw !== 'object' || raw === null) return false
  const s = raw as Partial<ChartSpec>
  if (typeof s.type !== 'string' || !KNOWN_TYPES.includes(s.type)) return false
  if (!Array.isArray(s.data) || s.data.length === 0) return false
  if (typeof s.x !== 'string' || s.x === '') return false
  if (!Array.isArray(s.series) || s.series.length === 0 || s.series.length > 2) return false
  const keys = [s.x, ...s.series.map((se) => se?.key)]
  if (keys.some((k) => typeof k !== 'string' || k === '')) return false
  // x 与每个 series.key 都要能在 data 行里找到（否则画不出来，属配错）。
  return keys.every((k) => s.data!.some((row) => row !== null && typeof row === 'object' && k in row))
}

File: app/schemas/chart.py
"""图表契约（ChartSpec）：投递页分析「自由配图」的稳定形状（2026-09-15 方向定稿）。

analyze Agent 在产出分析叙事的同时，**可选**地配 0~2 张图辅助说明——图是它的自由产出，
但形状必须收敛到这份契约，前端才敢渲染（配错图前端另有一道运行时降级，本层是「契约尽量
不让坏 spec 出门」）。

约束理由：
- **6 种图型**（bar/line/area/donut/pie/radar）：demo 验证过的分析构图全集，覆盖
  分布/对比/占比/趋势/多维画像。桑基/热力/漏斗等对数据形状苛刻，模型自由发挥易配废，不开放。
- **series ≤ 2**：「市场 vs 你」这种二元对比是上限，更多列就是花图。
- **x / series.key 必须在 data 行里真实存在**：防 LLM 编出对不上的列名（契约先拦一道）。

双侧同构：本文件（Python）与 `frontend/src/types/chart.ts`（TS）字段一一对应，改一边必须改另一边。
"""

from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

# 开放的 6 种图型（前端 @mantine/charts 渲染器一一对应）。
ChartType = Literal["bar", "line", "area", "donut", "pie", "radar"]


class ChartSeries(BaseModel):
    """一个数值列：key 指向 data 行的字段名；label/color 展示用（可空，前端给默认）。"""

    key: str
    label: str = ""
    color: str = ""


class ChartSpec(BaseModel):
    """一份声明式图表描述：analyze Agent 产出，前端纯渲染（不附带判断逻辑）。"""

    type: ChartType
    title: str = ""
    data: list[dict[str, Any]] = Field(min_length=1)
    x: str  # 类目/时间轴字段名
    series: list[ChartSeries] = Field(min_length=1, max_length=2)
    # bar 专用：horizontal=纵向柱（类目在 X）/ vertical=横向条（中文长标签好排）。
    orientation: Literal["horizontal", "vertical"] = "horizontal"
    note: str = ""  # 口径标注（如「仅猎聘样本」），诚实标注数据来源

    @model_validator(mode="after")
    def _keys_exist_in_data(self) -> "ChartSpec":
        """X 与每个 series.key 都必须能在 data 行里找到——否则图根本画不出来，属配错。"""
        keys = {self.x, *(s.key for s in self.series)}
        if not all(any(k in row for row in self.data) for k in keys):
            raise ValueError(f"chart spec 引用了 data 里不存在的字段：{sorted(keys)}")
        return self
